website/blog/2025-07-01-mlflow-prompt-evaluate/index.md

```diff
@@ -0,0 +1,435 @@
+                    { "type": "text", "text": "what's in this image?" },
+                    {
+                        "type": "image_url",
+                        "image_url": {
+                            "url": f"data:image/jpeg;base64,{inputs}",
+                        },
+                    },
+                ],
+            }
+        ],
+    )
+
+    return completion.choices[0].message.content
+
+with mlflow.start_run() as run:
+    predicted = get_completion(images[0])
+    print(predicted)
+
+```
+
+[WIP] Add a screenshot for the logged runs of the LLM completion
+
+#### 1. Defining a custom Gen AI metric for evaluation
+
+You can create a custom LLM-as-a-judge metric within MLflow using [mlflow.metrics.genai.make_genai_metric()](https://mlflow.org/docs/latest/api_reference/python_api/mlflow.metrics.html#mlflow.metrics.genai.make_genai_metric/). For simplicity, we define a (custom) metric called _correct_format_, which returns a boolean value depending on whether the output from the LLM contains a list of dicts with the required keys 'question' and 'answer'. The _definition_ provides an explanation for the new metric and the _grading_prompt_ lets you set the grading criteria, which the LLM can use to compute the metric.
+
+```python
+correct_format = mlflow.metrics.genai.make_genai_metric(
+    name="correct_format",
+    definition=(
+        """The answer is a list of dicts where keys are `question` and `answer`."""
+    ),
+    grading_prompt=(
+        """If formatted correctly, return 1. Otherwise, return 0."""
+    ),
+    model="openai:/gpt-4o-mini",
+    greater_is_better=True,
+)
+```
+
+#### 2. MLflow Evaluate
+
+In this section, we pass the defined metric _correct_format_ to `mlflow.evaluate()`, which makes for every image selected earlier. The JSONs created from the OCR extraction using `get_json()` can used as the ground truth
+
+```python
+def batch_completion(df: pd.DataFrame) -> list[str]:
+    return [get_completion(image) for image in df["inputs"]]
+
+eval_result = mlflow.evaluate(
+    model=batch_completion,
+    data=pd.DataFrame({"inputs": images, "truth": jsons}),
+    targets="truth",
+    model_type="text",
+    extra_metrics=[correct_format],
+)
+```
+
+[WIP] - Add a screenshot of the MLflow UI + explanation
+
+### Applying Prompt Registry and Evaluate to the OCR use case
+
+Let's assume you are developing an OCR application, with different roles contributing to the project development. Here's how a Prompt engineer and a ML engineer benefit from these MLflow capabilities
+
+#### Prompt Engineer: Improve the Prompt
+
+As a Prompt Engineer, you'll be iteratively modifying the prompts and `mlflow.genai.register_prompt()` lets you version different prompts/ system messages that you craft, with the provision to set a template text, and include prompt metadata like author, project, etc. It is also possible to compare different versions of the prompt, and also search for specific prompts from the past.
```

Comment:
Plz rewrite for clarity/grammar. 

"You are an expert at making techincal documentation. Rewrite the following text to be more clear, concise, and expand upon areas that need clarification. Maintain professional yet simple language." is a good starting prompt for GPT to rewrite this